### PR TITLE
[/MM-57259] db/migrations: add new migration to fix an error on migration 96

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
Fixes an error, an artifact remaining from the older migrations. This also needs to be cherry-picked to ESRs.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-57259

#### Release Note

```release-note
An artifact on Threads table is removed with a new migration. This will cause a table re-write.
```

### CODE REVIEW

File: server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql
Comment (lieut-data) [inline server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql:14]: I understand this will make the tables consistent, but this also means downtime during the upgrade. Would we consider leaving the column and considering it deprecated instead? Or if it's causing an issue, can we relax any non-null requirements so that the current application (which doesn't write to it) won't trigger these? Anything to avoid the table rewrite.

**Review by lieut-data — CHANGES_REQUESTED** [review]
Thanks, @isacikgoz! One thought below re: avoiding the table rewrite.

File: server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql
Comment (isacikgoz) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql:14]: @lieut-data this issue was discovered during a customer migration. We can't specifically ignore a column easily. What we can do is basically we can try relaxing this issue by a) entirely ignoring the table during the migration and creating views ([suggestion](https://stackoverflow.com/a/45690787) from the original author) b) dropping the column just before the migration with a `BEFORE LOAD` clause but that would be against of being migration to be safe. c) create the same column in postgres which would be more problematic IMHO.

Neither of these seemed efficient to me but I'm open for alternatives and further discussions.

Comment (agnivade) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql:14]: I agree with Jesse. We cannot make blocking migrations any more unfortunately. Both a or b sound like reasonable alternatives to me.

**Review by agnivade — CHANGES_REQUESTED** [review]
Requesting to look for other alternatives.

File: server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql
Comment (isacikgoz) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql:14]: Thanks for the feedback @lieut-data @agnivade what if we used `ALGORITHM=INSTANT`? We already have mysql-8 requirement anyway.

File: server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql
Comment (lieut-data) [inline server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql:6]: Can you elaborate on this part?

**Review by lieut-data — APPROVED** [review]

File: server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql
Comment (lieut-data) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql]: I'm wondering if we should skip this check and "let it fail" if we can't apply it instantly? Otherwise the user ends up with a working, but blocking migration (violating our guarantees).

Comment (isacikgoz) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql]: I also added this to the migration guide: https://docs.mattermost.com/deploy/postgres-migration.html#artifacts-may-remain-from-previous-configurations-versions cc @lieut-data

Comment (lieut-data) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql]: > It will only run if it's a non-blocking migration (which should be the ideal and expected scenario).

To clarify, if a customer had compressed row formats, would it be expected that some parts of the migration are silently skipped? I wonder if an error on failing to do something is a more ideal scenario?

Comment (isacikgoz) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql]: I see, we can log that separately for sure but the migration can't fail (it would stop application from starting). I agree that this is not an ideal scenario but also thinking whether we should allow using different row formats.

Comment (lieut-data) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql]: > the migration can't fail (it would stop application from starting)

Why can't the migration fail? If we can't migrate for some reason, is failure a preventative safeguard against starting the application with an unsupported migration?

Comment (isacikgoz) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql]: Yes, we don't have a graceful recovery other than rolling back to older version/schema.

Comment (lieut-data) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql]: > Yes, we don't have a graceful recovery other than rolling back to older version/schema.

Right, but that's the correct thing to do if a migration fails, right? I'm currently thinking that we should never allow a "silent failure".

**Comment by agnivade** [comment]
@isacikgoz - Please request a review once you are ready with the changes :P Otherwise I might have forgotten about it.

**Review by agnivade — APPROVED** [review]
Looks good!

File: server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql
Comment (isacikgoz) [reply server/channels/db/migrations/mysql/000121_drop_threads_teamid.up.sql]: I see, this is what the compromise we were thinking by simply ignoring the fact that column may exist. @lieut-data we can log this by adding a new value to the `Systems` table to indicate that this was skipped. I wouldn't log this somewhere else as it can get lost and would not be easy to figure out if it was skipped (ie silently failed). WDYT? Would it make sense to you?

**Review by lieut-data — APPROVED** [review]
Approving with comment thread above -- I'll defer to you @isacikgoz if the tradeoff here makes sense, just wanted to make sure we had all the context.

**Comment by isacikgoz** [comment]
Closing this as this specific case being handled by the migration-assist and also addressed by the docs page [here](https://docs.mattermost.com/deploy/postgres-migration.html#artifacts-may-remain-from-previous-configurations-versions).

**Pull request closed without merging**
